### ISSUE TITLE
Testing infrastructure for ReactorMixin tests: hooking up two protocols to each other

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5392]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5392|
|Type|enhancement|
|Created|2011-11-16 02:36:57Z|
|Branch|https://github.com/twisted/twisted/tree/connect-protocols-5392-4|

I've implemented the following code a few times: run one protocol on server, another protocol connecting to it as client, run the reactor until they both disconnect and then gathering the resulting disconnect reasons. There should be utility methods for easily doing this, and at least a couple of the existing implementations should be ported to use this as a demonstration that the API is generic enough.

<details><summary>Searchable metadata</summary>

```
trac-id__5392 5392
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__ 
branch__branches_connect_protocols_5392_4 branches/connect-protocols-5392-4
branch_author__itamar__washort itamar, washort
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1321411017000000 1321411017000000
changetime__1334200500000000 1334200500000000
version__None None
owner__itamar itamar
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [33177]) Branching to 'connect-protocols-5392'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

Ugh. `ConnectionTestMixins` does this in a half-assed, badly factored unhelpful sort of way. My code is better. 

Do we need backwards compatibility for `ConnectionTestMixins` or can I refactor it?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

My end goal, BTW, is to refactor both the abort tests and pretty much everything in test_tls to use the new utility method. But if `ConnectionTestMixins` should stay as is, I might stop after the abort tests.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Test packages aren't public.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

Ready for review. I've refactored some, but not all, tests in `twisted.internet.test.test_tcp`, `test_unix` and `test_tls` to use this new infrastructure, and it will also be useful for future work on reactor tests (e.g. a thorough half-close test suite as part of [#5341](https://github.com/twisted/twisted/issues/5341), or an update to [#5347](https://github.com/twisted/twisted/issues/5347) if this gets merged first).

I am mystified by the PortableGtkReactor failures (http://buildbot.twistedmatrix.com/builders/winxp32-py2.7/builds/622). Any ideas?

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/connect-protocols-5392

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

Ah. Because PortableGtkReactor is a broken piece of crap with minimum latency of 100ms between any network event loop iteration, and so (a) it's slow (b) it will fail tests with short timeouts. The easy fix is to make it have less crappy latency (which will also speed up the test suite). But perhaps that should be in a different ticket?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

  1. I found it a bit odd that you called the arguments `serverProtocolFactory` and `clientProtocolFactory`, despite the fact that they're not, well, protocol factories. I'd suggest renaming them to something like 
  2. the `PortableGtkReactor` failures do need to be addressed, since that's a supported platform.  File another ticket, and maybe I'll have a look at that so I can get credit for this review? ;-)

Otherwise it looks pretty good and "` 6 files changed, 270 insertions(+), 295 deletions(-)`" speaks for itself.  If `PortableGtkReactor` can be made to pass its tests, then this can land.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set owner to @itamarst|
|-|-|

Actually I guess I should move it out of review, since other people should look at other tickets first.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst commented|
|-|-|

(In [33445]) Branching to 'connect-protocols-5392-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst removed owner|
|-|-|

I renamed the parameters, and tests now pass on portable gtk reactor:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/connect-protocols-5392-2

Putting up for review again (which can be quick) because I additionally also refactored `test_newtls.py` to use the new code.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @exarkun|
|-|-|


  1. It doesn't look like the parameters were renamed.
  1. The proliferation of new timeouts with new values is bad.  Any test that involves real I/O is going to take an unpredictable amount of time.  If we are ever going to deal with it, we really need to avoid having twenty different timeouts spread out all over the place.  Beyond that, tests with 1 second timeouts are going to fail on the buildbots, a lot (The `abortConnection` tests with their custom 10 second timeout are already failing a lot).
  1. Why is this a `ReactorBuilder` method?  Making it so provides an extremely short-term convenience (lots of existing test methods can start using it with minimum typing) but makes it inaccessible to any other code and collides with any potential future kind of connection testing any of these tests might want to do (not catastrophically, but I can imagine a future with 5 different connectiong testing helpers, confusing people about which one they want to use).  The implementation doesn't really need to be part of `ReactorBuilder` - it could even accept a reactor as an argument and not need to be mixed in to a `ReactorBuilder`/`EndpointCreator` either.  The extra assertions about `removeAll()` and `getDelayedCalls()` seem orthogonal, too.
    1.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

ugh, accidental early submit.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @itamarst|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_14"></a>|@itamarst commented|
|-|-|

(In [33590]) Apparently I forgot to commit the renamed parameters change. Refs [#5392](https://github.com/twisted/twisted/issues/5392)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_15"></a>|@itamarst removed owner|
|-|-|

I moved `connectProtocols` out of the class and made it a function, and removed the timeout parameter.

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/connect-protocols-5392-2

(I think the API documentation buildslave failure is unrelated to this branch, since it talks about Agent.)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_16"></a>|@itamarst commented|
|-|-|

Getting this in would be helpful for testing [#3204](https://github.com/twisted/twisted/issues/3204).

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_17"></a>|@washort commented|
|-|-|

(In [33717]) Branching to connect-protocols-5392-3

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_18"></a>|@washort commented|
|-|-|

(In [33723]) Apparently I forgot to commit the renamed parameters change. Refs [#5392](https://github.com/twisted/twisted/issues/5392)

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_19"></a>|@washort commented|
|-|-|

Okay my use of `bzr rebase` was a little clumsy but I resolved a conflict with trunk and put it into `connect-protocols-5392-3`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun set owner to @itamarst<br>@exarkun set status to `new`|
|-|-|

Thanks!  I'm really happy to see this refatoring.  Overall I think it's a great change.  I see a few not-so-major things with the code, though:

  1.  in `twisted/internet/test/connectionmixins.py`
    1. There's some random whitespace change in the `serverFactoryFor` docstring
  1. in `twisted/internet/test/reactormixins.py`
    1. Potential error-handling shortcomings in the implementation of `connectProtocols`:
      1. `serverEndpoint.listen` Deferred could fail, the result is a hang I think
      1. Same goes for `clientEndpoint.connect`
    1. A 5-tuple is not a radical data structure.  Perhaps an `epsilon.structlike.record`?  Oh wait that's not part of Twisted yet.  Well, something with attributes instead of indexes would be nice.
  1. in `twisted/internet/test/test_tls.py`
    1. `WrapperFactory.buildProtocol`'s _self__ argument could probably be better named (I imagine someone coming along and deleting an apparently accidental stray _ and then being confused).  "wrapperSelf" or something like that would probably do nicely.
    1. Also, `WrapperFactory` is gross :(  I liked `StartTLSClientEndpoint` better before. :/  Maybe we can figure out a way to clean it up more later, though.
    1. `StartTLSCreator` only uses `startTLS` on the client side, so the name is a little misleading.
  1. Since this is private infrastructure, do we really want to include it in the news file?
  1. Also, what do you think of adjusting the name of this new helper function?  `connectProtocols` doesn't suggest anything about running a reactor or waiting for a full connection/session to complete.  unfortunately all of the other names that come to mind have "and" in them at least once, which isn't any better.

I also see some possible future improvements:

  1. Using `twisted.protocols.policies` to remove the need for subclassing to use `connectProtocols`.
  1. Allowing protocol instances to be passed in, instead of callable factories (simplifies more user code than it complexifies, I _think_)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_22"></a>|@itamarst removed owner|
|-|-|

I addressed all the review comments except WrapperFactory being gross. I also changed the API so protocol instances are passed in, which simplified things quite a bit. Started buildbot here:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/connect-protocols-5392-3

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_23"></a>|@itamarst commented|
|-|-|

(In [34147]) Branching to 'connect-protocols-5392-4'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_24"></a>|@itamarst commented|
|-|-|

OK, merged forward.

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/connect-protocols-5392-4

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun set owner to @itamarst<br>@exarkun set status to `new`|
|-|-|

Thanks!

  1. `twisted/internet/test/reactormixins.py`
    1. The docs for `ConnectableProtocol` should probably point out that two of its attributes get set by some outside code.  Perhaps it's worth changing this, even, and adding a method to `ConnectableProtocol` for setting these.
    1. The docstring also says `L{ReactorBuilder.runProtocolsWithReactor}`, no longer accurate.
    1. It looks like it would be easy to pull `Factory` out of `runProtocolsWithReactor`.
    1. `runProtocolsWithReactor` will fail if the endpoint `listen` call returns a synchronously failing Deferred (eg because a port couldn't be bound).  `finished` will run and call `reactor.stop()` before `reactor.run()` has happened.  `twisted.internet.test.connectionmixins.needsRunningReactor` was recently added to help handle this case in some other tests.
    1. The `isinstance` in `finished` can be avoided if `log.err` were to be added as a separate errback prior to the `reactor.stop()`-calling callback.
    1. Also, we should really, really stop adding `log.err` calls without a 2nd argument.
  1. `twisted/internet/test/test_tcp.py`
    1. `twisted.internet.test.test_tcp.AbortConnectionTestCase_Win32Reactor.test_resumeProducingThrows` seems to fail a lot on the Windows Python 2.6 builder.
    1. The `TCPCreator` class docstring has _C{ReactorBuilder.runProtocolsWithReactor}_.  This is the wrong name now, and it should probably be `L{}` instead.
    1. `TCP4ClientTestsBuilder` needs a class docstring
    1. `TCPClientTestsBuilder` and `TCP6ClientTestsBuilder` need _better_ class docstrings.
    1. I think `TCPClientTestsBuilder` is (still) misnamed.  It's not a builder in the same way as the other things in this file.  It is _intentionally_ not used to generate test cases with `makeTestCaseClasses`.
    1. Some (technically pre-existing) issues with `runAbortTest`:
      1. The comment above `getDelayedCalls()` has a mistake - it says `buildReactor`, but I'm pretty certain it means `runReactor`.  A way to avoid this special check would be to cancel the timeout at the end of `runReactor`, I think.  Maybe file a ticket for that?
      1. I think that the return value of `flushLoggedErrors` should basically always be checked, at least for its length.
  1. `twisted/internet/test/test_tls.py`
    1. `StartTLSClientEndpoint` would be improved by the switch protocol ticket, right?  Can you add a link to that ticket above `WrapperFactory`?
  1. I think there are a few more tests that can be converted to use this new API.  Can you identify one or two and file tickets for converting them?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_27"></a>|@itamarst removed owner|
|-|-|

  * 2.1 will be addressed by [#5393](https://github.com/twisted/twisted/issues/5393).
  * 2.6.1 will be addressed by [#5634](https://github.com/twisted/twisted/issues/5634).
  * 4 is addressed by [#5635](https://github.com/twisted/twisted/issues/5635). I also want to use this infrastructure in [#3204](https://github.com/twisted/twisted/issues/3204), and I also got the impression that this would supersede [#5622](https://github.com/twisted/twisted/issues/5622).

All other comments were addressed, I hope. I started another buildbot run:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/connect-protocols-5392-4

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun set owner to @itamarst<br>@exarkun set status to `new`|
|-|-|

Thanks!  Looks great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_30"></a>|@itamarst set status to `closed`|
|-|-|

(In [34184]) Merge connect-protocols-5392-4: Utility function for connecting two protocols.

Author: itamar, washort
Review: exarkun
Fixes: [#5392](https://github.com/twisted/twisted/issues/5392)

Add new utility method to reactormixins test infrastructure, runProtocolsWithReactor. This connects two protocols with two endpoints, allowing the simplification of many of our more convoluted reactor tests.
